limitador/src/storage/in_memory.rs

```diff
@@ -2,95 +2,78 @@ use crate::counter::Counter;
 use crate::limit::{Limit, Namespace};
 use crate::storage::atomic_expiring_value::AtomicExpiringValue;
 use crate::storage::{Authorization, CounterStorage, StorageErr};
+use moka::sync::Cache;
 use std::collections::hash_map::Entry;
 use std::collections::{HashMap, HashSet};
-use std::hash::{Hash, Hasher};
-use std::sync::RwLock;
+use std::ops::Deref;
+use std::sync::{Arc, RwLock};
 use std::time::{Duration, SystemTime};
 
-#[derive(Eq, Clone)]
-struct CounterKey {
-    set_variables: HashMap<String, String>,
-}
-
-impl CounterKey {
-    fn to_counter(&self, limit: &Limit) -> Counter {
-        Counter::new(limit.clone(), self.set_variables.clone())
-    }
-}
-
-impl From<&Counter> for CounterKey {
-    fn from(counter: &Counter) -> Self {
-        CounterKey {
-            set_variables: counter.set_variables().clone(),
-        }
-    }
-}
-
-impl From<&mut Counter> for CounterKey {
-    fn from(counter: &mut Counter) -> Self {
-        CounterKey {
-            set_variables: counter.set_variables().clone(),
-        }
-    }
-}
-
-impl Hash for CounterKey {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        self.set_variables.iter().for_each(|(k, v)| {
-            k.hash(state);
-            v.hash(state);
-        });
-    }
-}
-
-impl PartialEq for CounterKey {
-    fn eq(&self, other: &Self) -> bool {
-        self.set_variables == other.set_variables
-    }
-}
-
-type NamespacedLimitCounters<T> = HashMap<Namespace, HashMap<Limit, HashMap<CounterKey, T>>>;
+type NamespacedLimitCounters<T> = HashMap<Namespace, HashMap<Limit, T>>;
 
 pub struct InMemoryStorage {
     limits_for_namespace: RwLock<NamespacedLimitCounters<AtomicExpiringValue>>,
+    qualified_counters: Cache<Counter, Arc<AtomicExpiringValue>>,
 }
 
 impl CounterStorage for InMemoryStorage {
     fn is_within_limits(&self, counter: &Counter, delta: i64) -> Result<bool, StorageErr> {
         let limits_by_namespace = self.limits_for_namespace.read().unwrap();
 
         let mut value = 0;
-        if let Some(limits) = limits_by_namespace.get(counter.limit().namespace()) {
-            if let Some(counters) = limits.get(counter.limit()) {
-                if let Some(expiring_value) = counters.get(&counter.into()) {
-                    value = expiring_value.value();
-                }
+
+        if counter.is_qualified() {
+            if let Some(counter) = self.qualified_counters.get(counter) {
+                value = counter.value();
+            }
+        } else if let Some(limits) = limits_by_namespace.get(counter.limit().namespace()) {
+            if let Some(counter) = limits.get(counter.limit()) {
```

Comment:
what we are finding here is the AtomicExpiringValue, do you think it makes sense to be explicit and rename `counter` with something like `expiring_value` ? I also understand it's a _counter_ in the end